app/src/main/java/org/thoughtcrime/securesms/jobmanager/JobController.java

```diff
@@ -0,0 +1,509 @@
+
+    Log.i(TAG, JobLogger.format(job, "Scheduling a retry in " + delay + " ms."));
+    scheduler.schedule(delay, constraints);
+
+    notifyAll();
+  }
+
+  synchronized void onJobFinished(@NonNull Job job) {
+    runningJobs.remove(job.getId());
+  }
+
+  @WorkerThread
+  synchronized void onSuccess(@NonNull Job job, @Nullable Data outputData) {
+    if (outputData != null) {
+      List<JobSpec> updates = Stream.of(jobStorage.getDependencySpecsThatDependOnJob(job.getId()))
+                                    .map(DependencySpec::getJobId)
+                                    .map(jobStorage::getJobSpec)
+                                    .map(jobSpec -> mapToJobWithInputData(jobSpec, outputData))
+                                    .toList();
+
+      jobStorage.updateJobs(updates);
+    }
+
+    jobStorage.deleteJob(job.getId());
+    jobTracker.onStateChange(job, JobTracker.JobState.SUCCESS);
+    notifyAll();
+  }
+
+  /**
+   * @return The list of all dependent jobs that should also be failed.
+   */
+  @WorkerThread
+  synchronized @NonNull List<Job> onFailure(@NonNull Job job) {
+    List<Job> dependents = Stream.of(jobStorage.getDependencySpecsThatDependOnJob(job.getId()))
+                                 .map(DependencySpec::getJobId)
+                                 .map(jobStorage::getJobSpec)
+                                 .withoutNulls()
+                                 .map(jobSpec -> {
+                                   List<ConstraintSpec> constraintSpecs = jobStorage.getConstraintSpecs(jobSpec.getId());
+                                   return createJob(jobSpec, constraintSpecs);
+                                 })
+                                 .toList();
+
+    List<Job> all = new ArrayList<>(dependents.size() + 1);
+    all.add(job);
+    all.addAll(dependents);
+
+    jobStorage.deleteJobs(Stream.of(all).map(Job::getId).toList());
+    Stream.of(all).forEach(j -> jobTracker.onStateChange(j, JobTracker.JobState.FAILURE));
+
+    return dependents;
+  }
+
+  /**
+   * Retrieves the next job that is eligible for execution. To be 'eligible' means that the job:
+   *  - Has no dependencies
+   *  - Has no unmet constraints
+   *
+   * This method will block until a job is available.
+   * When the job returned from this method has been run, you must call {@link #onJobFinished(Job)}.
+   */
+  @WorkerThread
+  synchronized @NonNull Job pullNextEligibleJobForExecution(@NonNull JobPredicate predicate) {
+    try {
+      Job job;
+
+      while ((job = getNextEligibleJobForExecution(predicate)) == null) {
+        if (runningJobs.isEmpty()) {
+          debouncer.publish(callback::onEmpty);
+        }
+
+        wait();
+      }
+
+      jobStorage.updateJobRunningState(job.getId(), true);
+      runningJobs.put(job.getId(), job);
+      jobTracker.onStateChange(job, JobTracker.JobState.RUNNING);
+
+      return job;
+    } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)